File: others/900-1000/971. 翻转二叉树以匹配先序遍历.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *

MOD = int(1e9 + 7)
DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def flipMatchVoyage(self, root: Optional[TreeNode], voyage: List[int]) -> List[int]:
        ans = []
        can = True

        def dfs(node: Optional[TreeNode]) -> None:
            nonlocal i
            nonlocal can
            if not can:
                return

            if node.val != voyage[i]:
                can = False
                return
            i += 1
            if node.left and node.right:
                if node.left.val == voyage[i]:
                    dfs(node.left)
                    dfs(node.right)
                else:
                    ans.append(node.val)
                    dfs(node.right)
                    dfs(node.left)
                return
            if node.left:
                dfs(node.left)
            if node.right:
                dfs(node.right)

        i = 0
        dfs(root)
        return ans if can else [-1]

File: leetcode/900-1000/971. 翻转二叉树以匹配先序遍历.cpp
//
// Created by 123 on 25-6-3.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] =  {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
	vector<int> flipMatchVoyage(TreeNode* root, vector<int>& voyage) {
		vector<int> ans;
		int i = 0;
		bool flag = true;

		auto dfs = [&](this auto&& dfs, TreeNode* node) -> void {
			if (!node || !flag) return;

			// 检查当前节点值是否匹配
			if (node->val != voyage[i]) {
				flag = false;
				return;
			}
			i++;

			// 如果只有右子树或没有子树，直接递归
			if (!node->left) {
				dfs(node->right);
				return;
			}

			// 如果只有左子树，直接递归
			if (!node->right) {
				dfs(node->left);
				return;
			}

			// 两个子树都存在，判断是否需要翻转
			if (node->left->val == voyage[i]) {
				// 左子树匹配，不需要翻转
				dfs(node->left);
				dfs(node->right);
			} else if (node->right->val == voyage[i]) {
				// 右子树匹配，需要翻转
				ans.push_back(node->val);
				dfs(node->right);
				dfs(node->left);
			} else {
				// 左右子树都不匹配，无法完成
				flag = false;
				return;
			}
		};

		dfs(root);
		return flag ? ans : vector<int>({-1});
	}
};